src/execution/window_segment_tree.cpp

```diff
@@ -639,16 +621,35 @@ void WindowSegmentTree::Evaluate(WindowAggregatorState &lstate, const idx_t *beg
 	}
 	ltstate.FlushStates(false);
 
-	ltstate.Finalize(result, count);
+	// without EXCLUDE, finalize and write result right here
+	if (frame_part == FramePart::FULL) {
+		ltstate.Finalize(result, count, true);
+	}
 
 	//	Set the validity mask on  the invalid rows
-	auto &rmask = FlatVector::Validity(result);
-	for (idx_t rid = 0; rid < count; ++rid) {
-		const auto begin = begins[rid];
-		const auto end = ends[rid];
 
-		if (begin >= end) {
-			rmask.SetInvalid(rid);
+	auto &rmask = FlatVector::Validity(result);
```

Comment:
If you accumulate each range separately, then I think this code becomes a final pass. Yes, you have to recompute the bounds for each row, but that should be pretty cheap.